libamqpprox/amqpprox_httpauthintercept.cpp

```diff
@@ -0,0 +1,292 @@
+                                  stream,
+                                  response,
+                                  responseCb));
+}
+
+void HttpAuthIntercept::onRead(
+    std::shared_ptr<beast::flat_buffer>,  // Buffer must persists between reads
+    std::shared_ptr<beast::tcp_stream>                               stream,
+    std::shared_ptr<beast::http::response<beast::http::string_body>> response,
+    const ReceiveResponseCb &responseCb,
+    beast::error_code        ec,
+    std::size_t              bytes_transferred)
+{
+    boost::ignore_unused(bytes_transferred);
+
+    if (ec) {
+        const std::string errorMsg =
+            "Unable to receive http response from hostname " + d_hostname +
+            " on port " + d_port +
+            ", Error: {category: " + ec.category().name() +
+            ", message: " + ec.message() +
+            ", value: " + std::to_string(ec.value()) + "}";
+        LOG_ERROR << errorMsg;
+        authproto::AuthResponse errorResponseData;
+        errorResponseData.set_result(authproto::AuthResponse::DENY);
+        errorResponseData.set_reason(errorMsg);
+        responseCb(errorResponseData);
+        return;
+    }
+
+    authproto::AuthResponse authResponseData;
+    if (!authResponseData.ParseFromString(response->body())) {
+        const std::string errorMsg = "Unable to deserialize auth response "
+                                     "data received from http service.";
+        LOG_ERROR << errorMsg;
+        authproto::AuthResponse errorResponseData;
+        errorResponseData.set_result(authproto::AuthResponse::DENY);
+        errorResponseData.set_reason(errorMsg);
+        responseCb(errorResponseData);
+        return;
+    }
+    LOG_TRACE << "Response from auth route gate service at " << d_hostname
+              << ":" << d_port << d_target << ": "
+              << "[ Auth Result: "
+              << ((authResponseData.result() == authproto::AuthResponse::ALLOW)
+                      ? "ALLOW"
+                      : "DENY")
+              << ", Reason: " << authResponseData.reason()
+              << ((authResponseData.has_authdata())
+                      ? (", Auth mechanism: " +
+                         authResponseData.authdata().authmechanism())
+                      : "")
+              << " ]";
+    responseCb(authResponseData);
+
+    // Gracefully close the socket
+    stream->socket().shutdown(tcp::socket::shutdown_both, ec);
+
+    // not_connected happens sometimes so don't bother reporting it.
+    if (ec && ec != beast::errc::not_connected) {
+        const std::string errorMsg =
+            "Unable to close socket and shutdown connection gracefully for "
+            "auth route gate service at " +
+            d_hostname + ":" + d_port + d_target +
+            ", Error: {category: " + ec.category().name() +
+            ", message: " + ec.message() +
+            ", value: " + std::to_string(ec.value()) + "}";
+        LOG_WARN << errorMsg;
```

Comment:
Future thought: I wonder if we'll want to support a fail open configuration at some point

Reply:
Yes, I agree, we might need this, when we start supporting keep-alive feature.